### ISSUE TITLE
Avoiding mutable state?

### ISSUE THREAD

**Issue by aepsil0n** [sebcrozet/nphysics#26]
Hi, there. I was thinking about adopting `nphysics` as a base for a game engine. Now, I am currently looking into functional reactive game design and the current API of `nphysics` does not play well with that, as it heavily relies on mutable state.

Ideally I would like to have the physics engine transform a list of bodies into a new list of bodies each step. Maybe this could already be accomplished by having `World` implement `Clone` and then doing something like:

```rust
fn step(world: &World, dt: Scalar) -> World {
    let mut new_world = world.clone();
    new_world.step(dt);
    new_world
}
```

Thereby the mutability is confined to the local scope of this function, which would already be fine for my purposes. Of course, I hope that stepping typically takes more time than cloning, so that this is actually viable performance-wise. Do you think this is feasible?

**Comment by sebcrozet** [comment]
That sounds extremely hard − if not impossible − to do efficiently.

First, cloning the world would be extremely costly because you would have to clone all the acceleration structures too (including a bounding volume hierarchy).

On the other hand, not cloning the world but using it as a function that transforms some rigid bodies to a new list of rigid bodies (the original bodies are not mutated in-place, while the world itself is free to mutate its internal acceleration structures) might be feasible with a non-negligible overhead. At the very least, this implies that we copy each rigid body at each update. This is costly because a rigid body contains a *lot* of information (some of them being synchronized with some internal data related to acceleration structures). In addition, I believe that exploiting spacial/temporal coherence would be much harder because the World could not be fully aware of when a rigid body has been removed from the scene (the user being free to pass any list of rigid-body to the stepping function).

I did not think much about this, but that's my intuition. Also, a few years ago, I wrote a [purely functional physics engine](https://github.com/sebcrozet/falling) in haskell; it was a complete disaster performance-wise.
